### PR TITLE
[SR-4075] Reap dying QueryContext

### PR DESCRIPTION
**Pull request by satanson**

### CODE REVIEW

**Comment by kangkaisen** [comment]
Please describe what issue  this pr want to resolve，and how to resolve.

File: be/src/exec/pipeline/exec_state_reporter.h [hunk 1]
```diff
@@ -22,7 +22,7 @@ class ExecStateReporter {
                                                                     bool done);
     static Status report_exec_status(const TReportExecStatusParams& params, ExecEnv* exec_env, TNetworkAddress fe_addr);
     ExecStateReporter();
-    void submit(FragmentContext* fragment_ctx, const Status& status, bool done, bool clean);
+    std::unique_ptr<ThreadPoolToken> new_token() { return _thread_pool->new_token(ThreadPool::ExecutionMode::SERIAL); }
```
Comment (kangkaisen) [inline be/src/exec/pipeline/exec_state_reporter.h:25]: Comment why must use SERIAL `ExecutionMode`?

File: be/src/exec/pipeline/pipeline_driver_dispatcher.cpp [hunk 2]
```diff
@@ -30,14 +31,25 @@ void GlobalDriverDispatcher::change_num_threads(int32_t num_threads) {
 }
 
 void GlobalDriverDispatcher::run() {
+    // execution threads choose different strides to compute next slots to probe.
+    // different strides can reduce contention.
+    const auto slot_stride = _query_context_reaper->next_slot_stride();
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_driver_dispatcher.cpp:36]: Why introduce the `slot_stride` and `slot_size` conception ?

File: be/src/exec/pipeline/pipeline_driver_dispatcher.cpp [hunk 3]
```diff
@@ -30,14 +31,25 @@ void GlobalDriverDispatcher::change_num_threads(int32_t num_threads) {
...
+    const auto slot_size = _query_context_reaper->slot_size();
+    int slot_idx = 0;
     while (true) {
         if (_num_threads_setter.should_shrink()) {
             break;
         }
-
+        // reap in-active dying QueryContext
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_driver_dispatcher.cpp:43]: clean is more better than  reap？

File: be/src/exec/pipeline/pipeline_driver_queue.cpp [hunk 4]
```diff
@@ -44,7 +44,12 @@ DriverPtr QuerySharedDriverQueue::take(size_t* queue_index) {
                 break;
             }
             _is_empty = true;
-            _cv.wait(lock);
+            // wait for drivers' arrival, but awake execution threads to process dying QueryContexts after
+            // 100 micro-seconds' waiting.
+            auto cv_status = _cv.wait_for(lock, std::chrono::microseconds(100));
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_driver_queue.cpp:49]: This is reasonable？

File: be/src/exec/pipeline/exec_state_reporter.h (hunk 1 above)
Comment (liuyehcf) [reply be/src/exec/pipeline/exec_state_reporter.h:25]: The report tasks submitted in the same query need to ensure the order of execution

File: be/src/exec/pipeline/pipeline_driver_queue.cpp (hunk 4 above)
Comment (liuyehcf) [reply be/src/exec/pipeline/pipeline_driver_queue.cpp:49]: wake up and do clean job? 
I think it's fine to wake up till next pipeline arrive? (maybe unused objects will be last for a long time)

File: be/src/exec/pipeline/query_context.cpp [hunk 5]
```diff
@@ -43,10 +43,79 @@ QueryContextPtr QueryContextManager::get(const TUniqueId& query_id) {
...
+            577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701,
+            709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853,
+            857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997};
+    // binary search closest prime number:
+    // 1. n <= 997, return minimum prime number >= n.
+    // 2. n > 997, return 997.
+    int l = 0;
+    int h = sizeof(primes) / sizeof(primes[0]) - 1;
+    while (l < h) {
+        int m = (l + h) / 2;
```
Comment (liuyehcf) [inline be/src/exec/pipeline/query_context.cpp:79]: l - h >> 1 is faster

**Pull request closed without merging**